### PR TITLE
Pwpoly

### PR DESCRIPTION
**Pull request by max-collins**
## Description
Implementation of the PiecewisePolynomial class as well as relevant tests.
## Linked Issue
Fixes # (issue number)

## Implementation Checklist
- [x] My class inherits from `PolynomialBasis` (if applicable).
- [x] I have implemented all abstract methods (`evaluate_basis`, `fit`, etc.).
- [x] I have used `numpy` vectorization to avoid inefficient Python loops.
- [x] Documentation strings (docstrings) are provided for all public methods.

### 1. Mathematical Integrity
- [ ] **Convergence:** Does the implementation maintain expected accuracy? (Checked for Runge's phenomenon or stability).
- [ ] **Edge Cases:** Does it handle single points or non-equally spaced nodes correctly?
- [ ] **Domain Mapping:** Does the class correctly map the input domain to the basis domain (e.g., [-1, 1] for Chebyshev)?

### 2. API & Style
- [ ] **Method Signatures:** Methods match the Abstract Base Class exactly.
- [ ] **Naming:** Follows `CamelCase` for classes and `snake_case` for functions/variables.
- [ ] **Clean Code:** No "magic numbers"; constants are named and explained.

### 3. Testing Quality
- [ ] **Coverage:** Unit tests cover typical usage and edge cases.
- [ ] **Precision:** Tests use `np.testing.assert_allclose` instead of `==`.
- [ ] **Validation:** There is a test proving the basis can exactly reconstruct a polynomial of its own degree.

## Screenshots / Plots

### CODE REVIEW

**Comment by mtezzele** [comment]
@wdwatson2 please have a look at this PR as well

File: src/numanalysislib/basis/piecewise.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+import numpy as np
+import warnings
+
+
+class PiecewisePolynomial(PolynomialBasis):
+    def __init__(self, basis_type: PolynomialBasis, mesh: np.array):
+        """
+        :param basis_type: Type of Polynomial basis to use
+        :type basis_type: PolynomialBasis
+        :param mesh: Mesh points
```
Comment (wdwatson2) [inline src/numanalysislib/basis/piecewise.py]: More description would be helpful. Does it need to be increasing? Does it need at least two points?

File: src/numanalysislib/basis/piecewise.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+            # extract info from element
+            element = self.mesh[index]
+            a = element[0]
+            b = element[1]
+            n_dofs = self.bases[element].n_dofs
+
+            # create equidistant x_nodes to fit with
+            x_nodes = []
+            for i in range(n_dofs):
+                x_nodes.append((1-i/(n_dofs-1))*a + i/(n_dofs-1)*b)
```
Comment (wdwatson2) [inline src/numanalysislib/basis/piecewise.py:79]: what about when n_dofs = 1? This corresponds to constant poly right?

File: src/numanalysislib/basis/piecewise.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+            x_nodes = np.asarray(x_nodes)
+            #fit 
+            bases_coeffs[element] = self.bases[element].fit(x_nodes, y_element)
+        
+
+        return bases_coeffs
+    
+    def float_evaluate_basis(self, index: int, x: float) -> float:
```
Comment (wdwatson2) [inline src/numanalysislib/basis/piecewise.py]: docstring or do you intend this to be a private method?

File: src/numanalysislib/basis/piecewise.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+
+        :param index: index of basis function to evaluate
+        :type index: int
+        :param x: points to evaluate at
+        :type x: np.array
+        :rtype: np.array
+        """
+        return np.vectorize(lambda y: self.float_evaluate_basis(index, y))(x)
+
+    def float_evaluate(self, coefficients: dict, x: float) -> float:
```
Comment (wdwatson2) [inline src/numanalysislib/basis/piecewise.py]: same here

**Comment by wdwatson2** [comment]
your docstrings are probably formatted correctly and this might not matter but we were supposed to use numpy style (https://github.com/mtezzele/math-516-numerical-analysis-II-SP26/blob/main/.github/CONTRIBUTING.md)

File: src/numanalysislib/basis/piecewise.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+        # inheret attributes
+        self.basis = basis_type
+        self.a_hat = basis_type.a
+        self.b_hat = basis_type.b
+        self.degree = basis_type.degree
+        self.n_dofs = basis_type.n_dofs
+
+        self.mesh = []
+        for index in range(len(mesh) - 1):
+            self.mesh.append((mesh[index], mesh[index + 1]))
```
Comment (mtezzele) [inline src/numanalysislib/basis/piecewise.py]: this can be written in a more pythonic way:
`self.mesh = [(mesh[index], mesh[index + 1]) for index in range(len(mesh) - 1)]`

File: src/numanalysislib/basis/piecewise.py [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+            a = element[0]
+            b = element[1]
+            if x < a:
+                max_index = current_index
+            elif x > b:
+                min_index = current_index
+            else:
+                return self.bases[element].evaluate(coefficients[element], x) 
+            
+            current_index = min_index + (max_index - min_index)//2
```
Comment (mtezzele) [inline src/numanalysislib/basis/piecewise.py]: I think you can avoid the duplication of `current_index` by moving it as the first instruction in the for loop

File: src/numanalysislib/basis/piecewise.py [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+        for _ in range(len(self.mesh)):
+            element = self.mesh[current_index]
+            a = element[0]
+            b = element[1]
+            if x < a:
+                max_index = current_index
+            elif x > b:
+                min_index = current_index
+            else:
+                return self.bases[element].evaluate_basis(index, x)
```
Comment (mtezzele) [inline src/numanalysislib/basis/piecewise.py]: Please write a private method that does this loop so as to avoid the code duplication (see line 124+)
